File: src/lib/entities/Deck.ts
import PlayingCard from "$lib/entities/PlayingCard"
import PlayingCardSuite from "$lib/enums/PlayingCardSuit"
import PlayingCardValue from "$lib/enums/PlayingCardValue"

export default class Deck {
	private cards: PlayingCard[]

	constructor() {
		this.reset()
	}

	/**
	 * Put all cards back in the deck, in order
	 */
	public reset(): void {
		this.cards = []

		for (
			let suit = PlayingCardSuite.Spades;
			i <= PlayingCardSuite.Hearts;
			i++
		) {
			for (
				let value = PlayingCardValue.Ace;
				i <= PlayingCardValue.King;
				i++
			) {
				this.cards.push(new PlayingCard(suit, value))
			}
		}
	}

	/**
	 * Shuffle deck
	 */
	public shuffle(): void {
		for (let i = this.cards.length - 1; i > 0; i--) {
			const j = Math.floor(Math.random() * (i + 1))
			;[this.cards[i], this.cards[j]] = [this.cards[j], this.cards[i]]
		}
	}

	/**
	 * Draw a card, removing it from the deck,
	 * returns undefined if deck is empty
	 */
	public draw(): PlayingCard | undefined {
		return this.cards.pop()
	}

	/**
	 * Number of cards currently in the deck
	 */
	public get length(): number {
		return this.cards.length
	}
}

File: src/entities/deck.rs
use bevy::prelude::Resource;
use strum::IntoEnumIterator;
use crate::entities::PlayingCard;
use crate::enums::{PlayingCardSuit, PlayingCardValue};

#[derive(Resource)]
pub struct Deck {
	cards: Vec<PlayingCard>,
}

impl Deck {
	/// Maximum number of cards in deck
	pub const MAX: usize = 52;

	/// Put all cards back in the deck, in order
	pub fn reset(&mut self) {
		self.cards.clear();

		for suit in PlayingCardSuit::iter() {
			for value in PlayingCardValue::iter() {
				self.cards.push(PlayingCard::new(value, suit));
			}
		}
	}

	/// Shuffle deck
	pub fn shuffle(&mut self) {
		fastrand::shuffle(&mut self.cards);
	}

	/// Draw a card, removing it from the deck.
	/// Returns [`None`] if deck is empty
	pub fn draw(&mut self) -> Option<PlayingCard> {
		self.cards.pop()
	}

	/// Number of cards currently in the deck
	pub fn len(&self) -> usize {
		self.cards.len()
	}
}

impl Default for Deck {
	fn default() -> Self {
		Self {
			cards: Vec::with_capacity(Self::MAX),
		}
	}
}